pkg/proxy/winuserspace/proxier.go

```diff
@@ -304,322 +319,79 @@ func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
 
 		for i := range service.Spec.Ports {
 			servicePort := &service.Spec.Ports[i]
-			serviceName := proxy.ServicePortName{NamespacedName: types.NamespacedName{Namespace: service.Namespace, Name: service.Name}, Port: servicePort.Name}
-			activeServices[serviceName] = true
-			serviceIP := net.ParseIP(service.Spec.ClusterIP)
-			info, exists := proxier.getServiceInfo(serviceName)
-			// TODO: check health of the socket?  What if ProxyLoop exited?
-			if exists && sameConfig(info, service, servicePort) {
-				// Nothing changed.
-				continue
-			}
-			if exists {
-				glog.V(4).Infof("Something changed for service %q: stopping it", serviceName)
-				err := proxier.closePortal(serviceName, info)
-				if err != nil {
-					glog.Errorf("Failed to close portal for %q: %v", serviceName, err)
+			// create a slice of all the source IPs to use for service port portals
+			listenIPPortMap := getListenIPPortMap(service, int(servicePort.Port), int(servicePort.NodePort))
+			protocol := servicePort.Protocol
+
+			for listenIP, listenPort := range listenIPPortMap {
+				servicePortPortalName := ServicePortPortalName{NamespacedName: types.NamespacedName{Namespace: service.Namespace, Name: service.Name}, Port: servicePort.Name, PortalIPName: listenIP}
+				activeServicePortPortals[servicePortPortalName] = true
+				info, exists := proxier.getServiceInfo(servicePortPortalName)
+				if exists && sameConfig(info, service, protocol, listenPort) {
+					// Nothing changed.
+					continue
+				}
+				if exists {
+					glog.V(4).Infof("Something changed for service %q: stopping it", servicePortPortalName)
+					err := proxier.closeServicePortPortal(servicePortPortalName, info)
+					if err != nil {
+						glog.Errorf("Failed to close service port portal %q: %v", servicePortPortalName, err)
+					}
 				}
-				err = proxier.stopProxy(serviceName, info)
+				glog.V(1).Infof("Adding new service %q at %s:%d/%s", servicePortPortalName, listenIP, listenPort, protocol)
+				info, err := proxier.addServicePortPortal(servicePortPortalName, protocol, listenIP, listenPort, proxier.udpIdleTimeout)
 				if err != nil {
-					glog.Errorf("Failed to stop service %q: %v", serviceName, err)
+					glog.Errorf("Failed to start proxy for %q: %v", servicePortPortalName, err)
+					continue
 				}
+				info.sessionAffinityType = service.Spec.SessionAffinity
+				glog.V(4).Infof("info: %#v", info)
```

Comment:
This log feels spammy, if you think its worth keeping maybe increase the log level?